slapp/utils/merge_utils.py

```diff
@@ -0,0 +1,186 @@
+from typing import List, Union, Optional, TypedDict
+from pathlib import Path
+import copy
+import jsonlines
+import boto3
+from urllib.parse import urlparse
+import tempfile
+import shutil
+import logging
+import json
+
+from slapp.transfers.utils import read_jsonlines
+
+
+class MergingException(Exception):
+    pass
+
+
+class WorkerAnnotation(TypedDict):
+    workerId: str
+    roiLabel: str
+
+
+class RecordProject(TypedDict):
+    sourceData: str
+    majorityLabel: str
+    workerAnnotations: List[WorkerAnnotation]
+
+
+def compute_majority(labels):
+    """return majority labels or None if tied
+    """
+    ulabels = set(labels)
+    if len(ulabels) not in [1, 2]:
+        raise ValueError("`compute_majority()` expects binary classification "
+                         f"but received {len(ulabels)} labels: {ulabels}")
+    counts = {ulabel: labels.count(ulabel) for ulabel in ulabels}
+
+    majority = None
+
+    # require 3 and only 3 annotations
+    if sum(counts.values()) == 3:
+        mcount = max(counts.values())
+        r_counts = {v: k for k, v in counts.items()}
+        majority = r_counts[mcount]
+
+    return majority
+
+
+def get_record_project_key(record: dict):
```

Comment:
Perhaps `record` should also have a `TypedDict` entry?

Reply:
I wanted to do that, but here's the catch. The record keys `myproject` and `myproject-metadata` are different for every launched job. So, we don't know a-priori their names. Can you think of a solution?

Reply:
Maybe setting `total=False` and making a note about the undocumented `myproject` and `myproject-metadata` keys?

https://mypy.readthedocs.io/en/latest/more_types.html#totality

Reply:
right, but the hierarchy of this is:
```
Record - |
               | - RecordProject - |
                                             | - WorkerAnnotations
```
That's what would be useful, but for this:
```
wa = myrecord['myproject']['WorkerAnnotations']
```
`myproject` is dynamic, so I can't specify a TypedDict like:
```
class Record(TypedDict):
    myproject: RecordProject
```
because I don't know the name <myproject>

Reply:
I think a way to do this is :
```
class ParsedRecord(dict):
    @classmethod
    def from_record(cls, dict_in: dict):
        d = {}
        pkey = get_project_key(dict_in)
        if pkey is None:
            return None
        d['project-metadata'] = dict_in.pop(pkey + '-metadata')
        d['project'] = Project(dict_in.pop(pkey))
        d['project_key'] = pkey
        for k, v in dict_in.items():
            d[k] = v
        print(json.dumps(d, indent=2))
        return ParsedRecord(**d)

    def to_record(self):
        pkey = self.pop('project_key')
        self[pkey] = self.pop('project')
        self[pkey + '-metadata'] = self.pop('project-metadata')
```
but as we discussed offline, this is more work than we need right now. Maybe we can come back to this later if we want.